cmake/macros/macro_define_interface_target.cmake

```
@@ -0,0 +1,148 @@
+# defined then a single interface_<feature> target is defined.
+#
+#  - All interface targets are added automatically to
+#      ${DEAL_II_PROJECT_CONFIG_NAME}Targets.cmake
+#    for build and install locations.
+#
+# - The function also adds all targets to DEAL_II_TARGETS (for non
+#   debug/release split) and DEAL_II_TARGETS_(DEBUG|RELEASE) (for split
+#   targets), respectively.
+#
+# The default name, i.e., interface_<feature>(|_debug|_release) can be
+# overriden by the optional second argument. For example,
+#   define_interface_target(DEAL_II base_configuration)
+# will define interface_base_configuration* targets but query all
+# information from DEAL_II_* variables.
+#
+
+function(define_interface_target _feature)
+  string(TOLOWER "${_feature}" _feature_lowercase)
+  if(NOT "${ARGN}" STREQUAL "")
+    set(_feature_lowercase "${ARGN}")
+  endif()
+
+  set(_builds "dummy")
+  if(${_feature}_SPLIT_CONFIGURATION)
+    set(_builds ${DEAL_II_BUILD_TYPES})
+  endif()
+
+  foreach(_build ${_builds})
+    set(_interface_target "interface_${_feature_lowercase}")
+
+    if(${_feature}_SPLIT_CONFIGURATION)
+      string(TOLOWER "${_build}" _build_lowercase)
+      string(APPEND _interface_target "_${_build_lowercase}")
+    endif()
+
+    message(STATUS "")
+    message(STATUS "Defining target: ${_interface_target}")
+
+    add_library(${_interface_target} INTERFACE)
+
+    if(DEFINED ${_feature}_VERSION)
+      message(STATUS "    VERSION:             ${${_feature}_VERSION}")
+      set_target_properties(${_interface_target}
+        PROPERTIES VERSION "${${_feature_}_VERSION}"
+        )
+    endif()
+
+    set(_libraries)
+    list(APPEND _libraries
+      ${${_feature}_LIBRARIES} ${${_feature}_LIBRARIES_${_build}}
+      ${${_feature}_TARGETS} ${${_feature}_TARGETS_${_build}}
+      )
+    if(NOT "${_libraries}" STREQUAL "")
+      message(STATUS "    LINK_LIBRARIES:      ${_libraries}")
+      target_link_libraries(${_interface_target} INTERFACE ${_libraries})
+    endif()
+
+    if(DEFINED ${_feature}_INCLUDE_DIRS)
+      message(STATUS "    INCLUDE_DIRECTORIES: ${${_feature}_INCLUDE_DIRS}")
+      target_include_directories(${_interface_target}
+        SYSTEM INTERFACE ${${_feature}_INCLUDE_DIRS}
+        )
+    endif()
+
+    set(_definitions)
+    list(APPEND _definitions ${${_feature}_DEFINITIONS} ${${_feature}_DEFINITIONS_${_build}})
+    if(NOT "${_definitions}" STREQUAL "")
+      message(STATUS "    COMPILE_DEFINITIONS: ${_definitions}")
+      target_compile_definitions(${_interface_target} INTERFACE ${_definitions})
+    endif()
+
+    separate_arguments(_compile_options UNIX_COMMAND
+      "${${_feature}_CXX_FLAGS} ${${_feature}_CXX_FLAGS_${_build}}"
+      )
+    if(NOT "${_compile_options}" STREQUAL "")
+      message(STATUS "    COMPILE_OPTIONS:     ${_compile_options}")
+      target_compile_options(${_interface_target} INTERFACE $<$<COMPILE_LANGUAGE:CXX>:${_compile_options}>)
+    endif()
```

Can we just take the given compile options if they are defined (and CXX_FLAGS are not) so that we can avoid the dance we do for `Kokkos` (or any other package that has defined a proper `CMake` target) when parsing the flags already defined via `compile_options`? Something like
```suggestion
    if(${_feature}_compile_options)
      set(_compile_options "${${_feature}_compile_options}")
      message(STATUS "    COMPILE_OPTIONS:     ${_compile_options}")
    else()  
      separate_arguments(_compile_options UNIX_COMMAND
        "${${_feature}_CXX_FLAGS} ${${_feature}_CXX_FLAGS_${_build}}"
        )
      message(STATUS "    COMPILE_OPTIONS:     ${_compile_options}")
      set(_compile_options "$<$<COMPILE_LANGUAGE:CXX>:${_compile_options}>") 
    endif()  
    if(NOT "${_compile_options}" STREQUAL "")
      target_compile_options(${_interface_target} INTERFACE ${_compile_options})
    endif()
```
with the caveat of the status message being ugly in the first case. I haven't tested that.

---

@masterleinad My plan was to only support our legacy `FEATURE_*` variables.

For proper import targets (such as Kokkos) I plan to support a new `FEATURE_TARGETS(|_RELEASE|_DEBUG)` variable listing all import targets for a feature. Those will then be either used directly, or properly transferred by cloning target properties.

Meaning, the following signature should ultimately be sufficient:
```
process_feature(Kokkos
  TARGETS Kokkos::kokkos
  CLEAR [...]
  )
```

I have not yet pushed this code. I would like to keep it separate from this pull request.

---

Sounds good.